File: Python/convert-sorted-array-to-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param {integer[]} nums
    # @return {TreeNode}
    def sortedArrayToBST(self, nums):
        def BST(b, e):
            node = None
            if b<e:
                i = (b+e)/2
                node = TreeNode(nums[i])
                node.left = BST(b, i)
                node.right = BST(i+1, e)
            return node
            
        return BST(0, len(nums))

File: C++/convert-sorted-array-to-binary-search-tree.cpp
/**
 *  * Definition for a binary tree node.
 *   * struct TreeNode {
 *    *     int val;
 *     *     TreeNode *left;
 *      *     TreeNode *right;
 *       *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 *        * };
 *         */
class Solution {
private:
    TreeNode* BST(std::vector<int> &nums, int start, int end) {
        TreeNode* node = NULL;
        if(start<end) {
            int mid = (start+end)/2;
            node = new TreeNode(nums[mid]);
            node->left = BST(nums, start, mid);
            node->right = BST(nums, mid+1, end);
        }
        return node;
    }
public:
    TreeNode* sortedArrayToBST(std::vector<int> &nums) {
        return BST(nums, 0, nums.size());
    }
};
